File: android/aksara-core/src/main/kotlin/com/aksara/Flattener.kt
package com.aksara

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

/**
 * Flattens a nested i18next-style JSON object into a single `Map<String, String>`
 * keyed by dot-path, so every `t(...)` lookup is one O(1) hash hit instead of a
 * per-call tree walk.
 *
 * - Nested objects become dotted paths: `{"common":{"login":"…"}}` -> `common.login`.
 * - Arrays are indexed: `{"steps":["a","b"]}` -> `steps.0`, `steps.1`.
 * - Plural sub-keys are just nested keys: `common.items.one`, `common.items.other`.
 * - Numbers/bools are stringified; `null` is skipped.
 *
 * Mirror of the Swift `Flattener`.
 */
internal object Flattener {
    fun flatten(obj: JsonObject): Map<String, String> {
        val out = LinkedHashMap<String, String>(obj.size * 4)
        flatten(obj, "", out)
        return out
    }

    private fun flatten(element: JsonElement, prefix: String, out: MutableMap<String, String>) {
        when (element) {
            is JsonObject -> for ((key, child) in element) {
                val path = if (prefix.isEmpty()) key else "$prefix.$key"
                flatten(child, path, out)
            }
            is JsonArray -> element.forEachIndexed { index, child ->
                val path = if (prefix.isEmpty()) index.toString() else "$prefix.$index"
                flatten(child, path, out)
            }
            JsonNull -> Unit // skip nulls
            is JsonPrimitive -> out[prefix] = element.content
        }
    }
}

File: swift/Sources/Aksara/Flattener.swift
import Foundation

/// Flattens a nested i18next-style JSON object into a single `[String: String]`
/// keyed by dot-path, so every `t(...)` lookup is one O(1) hash hit instead of a
/// per-call tree walk.
///
/// - Nested objects become dotted paths: `{"common":{"login":"…"}}` → `common.login`.
/// - Arrays are indexed: `{"steps":["a","b"]}` → `steps.0`, `steps.1`.
/// - Plural sub-keys are just nested keys: `common.items.one`, `common.items.other`.
///   Plural *selection* happens at lookup time (see `PluralResolver`).
/// - Numbers/bools are stringified so a stray non-string value can't blank a screen.
/// - `null` and unsupported values are dropped (they never reach here in practice
///   because `SchemaValidator` rejects them first).
enum Flattener {
    static func flatten(_ object: [String: Any]) -> [String: String] {
        var out: [String: String] = [:]
        out.reserveCapacity(object.count * 4)
        flatten(object, prefix: "", into: &out)
        return out
    }

    private static func flatten(_ value: Any, prefix: String, into out: inout [String: String]) {
        switch value {
        case let dict as [String: Any]:
            for (key, child) in dict {
                let path = prefix.isEmpty ? key : prefix + "." + key
                flatten(child, prefix: path, into: &out)
            }
        case let array as [Any]:
            for (index, child) in array.enumerated() {
                let path = prefix.isEmpty ? String(index) : prefix + "." + String(index)
                flatten(child, prefix: path, into: &out)
            }
        case let string as String:
            out[prefix] = string
        case let number as NSNumber:
            out[prefix] = number.stringValue
        default:
            break // null / unsupported — skip
        }
    }
}
